cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageAdmin.java

```diff
@@ -217,8 +259,80 @@ private Lineage doComputeLineage(final NamespacedEntityId sourceData, long start
       }
     }
 
-    Lineage lineage = new Lineage(
-      Iterables.concat(Maps.transformValues(relations.asMap(), COLLAPSE_UNKNOWN_TYPE_FUNCTION).values()));
+    Lineage lineage;
+    if (rollup.contains("workflow")) {
+      // Get workflow Run IDs for all the programs in the relations
+      final Set<String> workflowIDs = new HashSet<>();
+      for (RelationKey key : relations.keySet()) {
+        Collection<Relation> allRelations = relations.get(key);
+        for (Relation relation : allRelations) {
+          ProgramId program = relation.getProgram();
+          RunRecordMeta runRecord = store.getRun(program, relation.getRun().getId());
```

Comment:
This is actually quite inefficient. Each call to `getRun` results in a new transaction and a HBase lookup. We should have the store supports a multi-get-runs method (e.g. `Map<ProgramRunId, RunRecordMeta> getRuns(Iterable<ProgramRunId> programRunIds)`)